enh: Add support for SO_PRIORITY socket option.

olerem opened the pull request:
This functionality is needed to test mutliqueue support of Ethernet drivers.

Here is one use case example how this functionality can be tested:
```
On the server side:
# create mqprio queue discipline on lan2 interface
tc qdisc add dev lan2 parent root handle 100 mqprio num_tc 4 map 0 1 2 3  queues 1@0 1@1 1@2 1@3  hw 0
iperf3 -s &
# collect queue stats
tc -s qdisc show dev lan2

On the client side:
# start tcp reverse stream with socket priority 2
iperf3 -c  172.17.0.1 -b100M  -l1472 -t100 -R --sock-prio 2

On the server side:
# collect queue stats
tc -s qdisc show dev lan2

qdisc mqprio 100: root tc 4 map 0 1 2 3 0 0 0 0 0 0 0 0 0 0 0 0 
             queues:(0:0) (1:1) (2:2) (3:3) 
 Sent 35748478 bytes 23639 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0
qdisc fq_codel 0: parent 100:4 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64 
 Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent 100:3 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64 
 **Sent 17628922 bytes 11649 pkt** (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent 100:2 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64 
 Sent 0 bytes 0 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
qdisc fq_codel 0: parent 100:1 limit 10240p flows 1024 quantum 1514 target 5ms interval 100ms memory_limit 32Mb ecn drop_batch 64 
 Sent 18119556 bytes 11990 pkt (dropped 0, overlimits 0 requeues 0) 
 backlog 0b 0p requeues 0
  maxpacket 0 drop_overlimit 0 new_flow_count 0 ecn_mark 0
  new_flows_len 0 old_flows_len 0
```

At this point we should be able to see increasing counters on "qdisc fq_codel 0: parent 100:3 ..."

olerem committed:
enh: Add support for SO_PRIORITY socket option.

This functionality is needed to test multi queue support of Ethernet
drivers.

Co-developed-by: Oleksij Rempel <o.rempel@pengutronix.de>
Signed-off-by: Oleksij Rempel <o.rempel@pengutronix.de>
Co-developed-by: Marc Kleine-Budde <mkl@pengutronix.de>
Signed-off-by: Marc Kleine-Budde <mkl@pengutronix.de>
---
changes v7:
- rebase against latest git (20.01.2026)
changes v6:
- Corrected the socket priority validation range from 0-15 to the documented
  0-6.
- Fixed a regression in JSON output where an explicit socket priority of 0 was
  incorrectly skipped due to using a boolean check on the new -1 default.
- Refactored argument parsing and socket option setting to correctly handle
  the new -1 default for "not set".
- Added missing error strings for IEBADSOPRIO and IESETSOPRIO to the error
  list.
- Removed redundant platform-specific text ("only supported on Linux") from
  documentation.
- Rebase on top of v3.20, make some rebase related changes

olerem commented:
json variable is lest. Will send new version soon.

olerem commented:
Hi @bmah888, can you please give me some feedback for this PR :)

olerem commented:
Hi @swlars, can you please give me some feedback for this PR :)

olerem commented:
Hi all, just a friendly reminder about this PR. Please let me know if you need any changes or additional context.

davidBar-On commented on src/iperf_api.c:
```diff
@@ -4613,6 +4648,16 @@ iperf_common_sockopts(struct iperf_test *test, int s)
     }
+
+#if defined(HAVE_SO_PRIORITY)
+    if ((opt = test->settings->sock_prio)) {
```
This may be a non-issue, this statement means that if the user set the priority parameter to 0, `sectsockopt()` will not be executed.  Although 0 is the default priority, the user may want for some reason that the `setsockopt()` will be executed with priority 0.  Therefore, it may be better that `sock_prio` will be set to `-1` (note that if this is done then it has to be taken into account in changes like the `cJSON_AddItemToObject()` change in `iperf_on_test_start()`).

davidBar-On commented on src/iperf_locale.c:
```diff
@@ -196,6 +196,9 @@ const char usage_longstr[] = "Usage: iperf3 [-s|-c host] [options]\n"
                            "                            Numeric values can be specified in decimal,\n"
                            "                            octal and hex (see --tos above).\n"
+#if defined(HAVE_SO_PRIORITY)
+                           "      --sock-prio N         set the socket priority (only supported on Linux)\n"
```
The `(only supported on Linux)` is redundant (and may even be incorrect), as the option will not be shown in OSs that does not support it because of the `#if defined`.

olerem committed:
Decouple 'test_start' JSON generation in iperf_on_test_start

Replace the monolithic iperf_json_printf() call for generating the
'test_start' JSON object with direct cJSON_Add... calls.

This refactoring removes a significant maintenance overhead, as it
eliminates the need to rewrite the entire format string whenever a
new, particularly conditional, parameter (such as sock_prio) needs to
be added to the test parameters block.

Signed-off-by: Oleksij Rempel <o.rempel@pengutronix.de>

olerem commented:
Hi @davidBar-On , I hope all issues are addressed now.

jtconsol commented:
This feature would be a very worthwhile addition to iperf!

@olerem Out of curiosity: Would it be difficult to support the `--sock-prio` option for server mode as well? Although it seems that I can just use reverse mode to the same effect.
